ratis-server/src/main/java/org/apache/ratis/server/storage/StorageImplUtils.java

```diff
@@ -17,38 +17,157 @@
+import org.apache.ratis.util.SizeInBytes;
 
 import java.io.File;
 import java.io.IOException;
-import java.util.concurrent.TimeUnit;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.apache.ratis.server.RaftServer.Division.LOG;
 
 public final class StorageImplUtils {
+  private static final File[] EMPTY_FILE_ARRAY = {};
 
   private StorageImplUtils() {
     //Never constructed
   }
 
+  public static SnapshotManager newSnapshotManager(RaftPeerId id) {
+    return new SnapshotManager(id);
+  }
+
   /** Create a {@link RaftStorageImpl}. */
-  public static RaftStorageImpl newRaftStorage(File dir, RaftServerConfigKeys.Log.CorruptionPolicy logCorruptionPolicy,
-      RaftStorage.StartupOption option, long storageFeeSpaceMin) throws IOException {
-    RaftStorage.LOG.debug("newRaftStorage: {}, {}, {}, {}",dir, logCorruptionPolicy, option, storageFeeSpaceMin);
+  public static RaftStorageImpl newRaftStorage(File dir, SizeInBytes freeSpaceMin,
+      RaftStorage.StartupOption option, Log.CorruptionPolicy logCorruptionPolicy) {
+    return new RaftStorageImpl(dir, freeSpaceMin, option, logCorruptionPolicy);
+  }
+
+  private static List<File> getExistingStorageSubs(List<File> volumes, String targetSubDir,
+      Map<File, Integer> dirsPerVol) {
+    return volumes.stream().flatMap(volume -> {
+          final File[] dirs = Optional.ofNullable(volume.listFiles()).orElse(EMPTY_FILE_ARRAY);
+          Optional.ofNullable(dirsPerVol).ifPresent(map -> map.put(volume, dirs.length));
+          return Arrays.stream(dirs);
+        }).filter(dir -> targetSubDir.equals(dir.getName()))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Get a list of existing subdirectories matching the given storage directory name from the given root directories.
+   */
+  static List<File> getExistingStorageSubs(String storageDirName, StartupOption option,
+      List<File> rootDirs, Map<File, Integer> dirsPerVol) throws IOException {
+    Preconditions.assertEmpty(dirsPerVol, "dirsPerVol");
+    final List<File> existingSubs = getExistingStorageSubs(rootDirs, storageDirName, dirsPerVol);
+    final int size = existingSubs.size();
+    if (option == StartupOption.RECOVER) {
+      if (size > 1) {
+        throw new IOException("Failed to " + option + ": More than one existing directories found " + existingSubs
+            + " for " + storageDirName);
+      } else if (size == 0) {
+        throw new IOException("Failed to " + option + ": Storage directory not found for " + storageDirName
+            + " from " + rootDirs);
+      }
+    } else if (option == StartupOption.FORMAT) {
+      if (size > 0) {
+        throw new IOException("Failed to " + option + ": One or more existing directories found " + existingSubs
+            + " for " + storageDirName);
+      }
```

Comment:
Can we move these checks inside `recoverExistingStorageDir` and `formatNewStorageDir`?

Reply:
@lokeshj1703 , thanks for the suggestion!  If these checks are moved inside, then it needs to pass the variables for the error messages.  Let's create a class so that the variables will be available.